### ISSUE TITLE
Feedback on this experimentation

### ISSUE THREAD

**Issue by oliviertassinari** [oliviertassinari/react-with-styles-interface-jss#1]
This project is an experimentation with the API proposed by [react-with-styles](https://github.com/airbnb/react-with-styles) and [JSS](https://github.com/cssinjs/jss).
The implementation is as simple as possible.

Some points I have noticed:
1. Switching the interface from _Aphrodite_ to _JSS_ isn't straightforward.
   There are some inconsistencies. E.g:
   1. `:hover` -> `&:hover`
   2. No global _CSS_ selector for _Aphrodite_ (`html` / `body` reset). 
   3. Inversed media query structure (https://github.com/cssinjs/jss/issues/246)
   
   I'm wondering how those inconsistencies could be resolved:
   1. The interface could fill the gap
   2. Those lib authors could agree on a common style API.
2. _JSS_ and _Aphrodite_ don't perform identically. I would need to try it on a larger project to have more realistic benchmark than on the `/docs` folder of this repository. I would also need to run the test using `NODE_ENV='production'`.
   - _JSS_:
     ![capture d ecran 2016-09-11 a 13 40 20](https://cloud.githubusercontent.com/assets/3165635/18417102/3b9b06c8-7827-11e6-918c-e223710079de.png)
   - _Aphrodite_:
     ![capture d ecran 2016-09-11 a 13 40 34](https://cloud.githubusercontent.com/assets/3165635/18417096/143a1d12-7827-11e6-9108-49a80943c6ed.png)
3. The _react-with-styles_ configuration seems to be using a singleton. That can be problematic when shipping a library of components using it, like [Material-UI](https://github.com/callemall/material-ui). Users may want to use a different configuration / interface. 
4. I can't find a clear API on how to handle server-side rendering from a _react-with-style_ point of view.
5. The _react-with-style_ API hide both `style` and `className` under an obscure object spread. That implies an overhead for _JSS_ and _react-native_ that only use `className` and `style` respectively. We could be more efficient by avoiding this layer of abstraction: unnecessary memory allocation and destructuration.
6. The _react-with-style_ API lack one `options` parameter for the `create` method.
   There are some quick nice [options available](https://github.com/cssinjs/jss/blob/master/docs/js-api.md#create-style-sheet-with-namespaces-enabled) with _JSS_.
   One workaround is to use a reserved property, like `options` when using the `withStyles` API.
7. I think that it would be better to wait for the component to be actually rendered to attach the stylesheet into the DOM, as [jss-theme-reactor](https://github.com/nathanmarks/jss-theme-reactor) is doing it. That would allow us to inline only the used CSS when doing server-side rendering 🚀 .
### TL;DR

The _react-with-styles_ approach is really interesting. 

From a _Material-UI_ point of view. I and @nathanmarks are not sold. I think that we should keep exploring [jss-theme-reactor](https://github.com/nathanmarks/jss-theme-reactor).

**Comment by kof** [comment]
1. Different JSON DSL's. This will be hard to achieve as it would require a standard, where there is no. Also a standard would mean we can't innovate  in terms of  a syntax. For e.g. we are currently working on this https://github.com/cssinjs/jss-expand. I see no way how we can unify this.
2. Benchmarking is very hard in this case, especially because aphrodite is using "asap" and this delays the rendering, basically you are measuring the air if you compare aphrodites async rendering with jss. Also check out [aphrodisiac](https://github.com/cssinjs/aphrodisiac)

**Comment by lencioni** [comment]
Thanks for putting this together!

> Switching the interface from Aphrodite to JSS isn't straightforward.

The quickest solution seems to be for for the interface to fill the gap--although I'm not sure what the performance consequences would be. You could probably make a Babel transform that would work though. It would be really interesting to see everyone work together on standardizing an API. I think that would be valuable even if it doesn't cover everything and even if some libraries decide to deviate in some cases. 

> The react-with-styles configuration seems to be using a singleton.

That's right. I believe that this was done to help resolve some bundle splitting issues we were having and we kept it in when extracting this code to its own project. It's quite possible that this doesn't actually make sense for it to be in this project though. Although I don't think I fully understand your concern here--can you flesh it out a little for me?

> I can't find a clear API on how to handle server-side rendering from a react-with-style point of view.

This is currently handled outside of react-with-styles. E.g. if you are using Hypernova and Aphrodite (via react-with-styles), you would still use [hypernova-aphrodite](https://github.com/airbnb/hypernova-aphrodite). It might be interesting to incorporate server rendering into the interfaces, but the benefit there seems to be pretty small in comparison (so I'm not really sure how worthwhile this would be). I'm certainly open to the idea though.

> The react-with-style API hide both style and className under an obscure object spread. That implies an overhead for JSS and react-native that only use className and style respectively. 

I don't think I quite understand what you mean by this. In the Aphrodite interface, styles that go through the `withStyles` call end up as `className` and POJOs that are passed directly to `css()` end up as `style`.

> We could be more efficient by avoiding this layer of abstraction: unnecessary memory allocation and destructuration.

It would be great to have a babel plugin to perform an optimization on this.

> The react-with-style API lack one options parameter for the create method. There are some quick nice options available with JSS. One workaround is to use a reserved property, like options when using the withStyles API.

I think this might be okay for some of the JSS options, like `named`, but would be riskier for others, like `media`. Any divergence reduces the utility of an abstraction layer so I'm a little cautious about this.

> I think that it would be better to wait for the component to be actually rendered to attach the stylesheet into the DOM

I believe this is how Aphrodite is working, FWIW.

**Comment by oliviertassinari** [comment]
@lencioni @kof Thanks for your feedback too.

@kof Thanks for sharing the [aphrodisiac](https://github.com/cssinjs/aphrodisiac/blob/master/src/index.js) project. That could definitely help  improving the current implementation!

> Benchmarking is very hard in this case

I'm measuring the activity between the first received byte to the last screen flushed by the GPU.
That's pretty much an e2e benchmarking. I'm not sure the _asap_ has anything to do with the result.

> The quickest solution seems to be for the interface to fill the gap

@lencioni Right, I see three approaches that we could use to fill the gap:
- Using the API: we can abstract some property away with the `theme` object. E.g. This is what I'm doing with the [`:hover`/`&:hover`](https://github.com/oliviertassinari/react-with-styles-interface-jss/blob/master/docs/src/withStyles.js#L14) property on the docs folder.
- Using the compilation time. E.g. We could have a Babel transformation plugin to inverse the `@media` query structure at the compilation time.
- Using the runtime.

However, @kof point is still right:

> it would require a standard,

What should be the standard?

> Although I don't think I fully understand your concern here--can you flesh it out a little for me?

My concern with the singleton configuration is regarding a library perspective. What if a library is using `react-with-style` using **interface A** and a user wants to use this library with `react-with-style` and a **interface B**?

> It might be interesting to incorporate server rendering into the interfaces, but the benefit there seems to be pretty small in comparison

Yes, the server-side rendering API is most likely going to be used only once. The migration seems straightforward:
- [JSS](https://github.com/cssinjs/examples/blob/gh-pages/react-ssr/src/server.js#L16)
- [Aphrodite](https://github.com/airbnb/hypernova-aphrodite/blob/master/src/index.js#L10)

> In the Aphrodite interface, styles that go through the withStyles call end up as className and POJOs that are passed directly to css() end up as style.

My point is the JSS interface is only using the `className` property. The other interfaces I'm aware of are using, at most, two properties. Spreading the `css()` method result to the components looks extreme.
However, I don't have a better option in mind. That looks like the biggest cost of this `react-with-style` abstraction.

> It would be great to have a Babel plugin to perform an optimization on this.

That would add some magic behind the scene. But I think that it would work 👍 .

> I believe this is how Aphrodite is working, FWIW.

It's my interface that is attaching all the styles to the DOM at the declaration time, not JSS.
I can use the _aphrodisiac_ approach or _jss-theme-reactor_ approach for improving this point.
